lib/vector-config-macros/src/configurable.rs

```diff
@@ -0,0 +1,765 @@
+        ::vector_config::schema::finalize_schema(schema_gen, &mut subschema, #field_metadata_ref);
+    }
+}
+
+fn generate_named_struct_field(
+    container: &Container<'_>,
+    field: &Field<'_>,
+) -> proc_macro2::TokenStream {
+    let field_name = field
+        .ident()
+        .expect("named struct fields must always have an ident");
+    let field_as_configurable = get_field_type_as_configurable(field);
+    let field_already_contained = format!(
+        "schema properties already contained entry for `{}`, this should not occur",
+        field_name
+    );
+    let field_key = field.name();
+
+    let field_schema = generate_struct_field(field);
+
+    // If there is no default value specified for either the field itself, or the container the
+    // field is a part of, then we consider it required unless the field type itself is inherently
+    // optional, such as being `Option<T>`.
+    let maybe_field_required =
+        if container.default_value().is_none() && field.default_value().is_none() {
+            Some(quote! {
+                if !#field_as_configurable::is_optional() {
+                    if !required.insert(#field_key.to_string()) {
+                        panic!(#field_already_contained);
+                    }
+                }
+            })
+        } else {
+            None
+        };
+
+    quote! {
+        {
+            #field_schema
+
+            if let Some(_) = properties.insert(#field_key.to_string(), subschema) {
+                panic!(#field_already_contained);
+            }
+
+            #maybe_field_required
+        }
+    }
+}
+
+fn generate_tuple_struct_field(field: &Field<'_>) -> proc_macro2::TokenStream {
+    let field_schema = generate_struct_field(field);
+
+    quote! {
+        {
+            #field_schema
+            subschemas.push(subschema);
+        }
+    }
+}
+
+fn build_named_struct_generate_schema_fn(
+    container: &Container<'_>,
+    fields: &[Field<'_>],
+) -> proc_macro2::TokenStream {
+    let (clt, _) = get_configurable_lifetime();
+
+    let mapped_fields = fields
+        .iter()
+        // Don't map this field if it's marked to be skipped for both serialization and deserialization.
+        .filter(|field| field.visible())
+        .map(|field| generate_named_struct_field(container, field));
+
+    quote! {
+        fn generate_schema(schema_gen: &mut ::schemars::gen::SchemaGenerator, overrides: ::vector_config::Metadata<#clt, Self>) -> ::schemars::schema::SchemaObject {
+            let mut properties = ::indexmap::IndexMap::new();
+            let mut required = ::std::collections::BTreeSet::new();
+
+            let metadata = Self::metadata().merge(overrides);
+            #(#mapped_fields)*
+
+            // TODO: We need to figure out if we actually use `#[serde(flatten)]` anywhere in order
+            // to capture not-specifically-named fields i.e. collecting all remaining/unknown fields
+            // in a hashmap.
```

Comment:
I'm pretty confident that we don't do that. There are only a couple places I'm aware of where we collect arbitrary fields and they're things like `librdkafka_options.*`, where it's explicitly namespaced.

Reply:
Yeah, that was also my gut feeling.

Reply:
Nit: should we remove these TODOs given the discussion here?

Reply:
I plan on removing them as part of the actual integration PR, because I believe that's when we'll be able to definitively answer these questions.

Reply:
Aha, fair enough 😄